pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SparkRBackend.scala

```diff
@@ -85,13 +87,35 @@ object SparkRBackend {
     try {
       // bind to random port
       val boundPort = sparkRBackend.init()
+      val serverSocket = new ServerSocket(0, 1)
+      val listenPort = serverSocket.getLocalPort()
+
       // tell the R process via temporary file
       val path = args(0)
       val f = new File(path + ".tmp")
       val dos = new DataOutputStream(new FileOutputStream(f))
       dos.writeInt(boundPort)
+      dos.writeInt(listenPort)
       dos.close()
       f.renameTo(new File(path))
+
+      // wait for the end of stdin, then exit
+      new Thread("wait for stdin") {
+        setDaemon(true)
+        override def run(): Unit = {
+          // any un-catched exception will also shutdown JVM
+          val buf = new Array[Byte](1024)
+          // shutdown JVM if R does not connect back in 10 seconds
+          serverSocket.setSoTimeout(10000)
+          val inSocket = serverSocket.accept()
+          serverSocket.close()
+          // wait for the end of socket, closed if R process die
+          inSocket.getInputStream().read(buf)
```

Comment:
In case of socket is closed, the `read()` will return -1 without IOException. 

`sparkRBackend.close()` could be called in main thread, then it exit. Because this thread is daemoned, so it will not block main thread from exiting.

Once the JVM exited, OS will cleanup all the resources (sockets, threads).